File: service-two/src/main/java/com/dlt/kafkadlt/service/ConsumerService.java
package com.dlt.kafkadlt.service;

import com.dlt.kafkadlt.model.Info;
import com.dlt.kafkadlt.model.Thing;
import lombok.extern.slf4j.Slf4j;
import org.springframework.kafka.annotation.KafkaListener;
import org.springframework.kafka.annotation.RetryableTopic;
import org.springframework.retry.annotation.Backoff;
import org.springframework.stereotype.Service;

import static com.dlt.kafkadlt.utils.Constants.*;

@Slf4j
@Service
public class ConsumerService {

    //    @KafkaListener(topics = TOPIC_GOOD, groupId = "group_id", errorHandler = "notificationListenerErrorHandler")
    @KafkaListener(topics = NORMAL_TOPIC)
    public void consumeNormalTopic(Thing message) {
        log.info("Got message {}", message);
    }

    @KafkaListener(topics = NORMAL_TOPIC, groupId = "third-group")
    public void consumeNormalTopicFail(Thing message) {
        log.info("Got message {}", message);
        throw new RuntimeException("Bad message");
    }

    @RetryableTopic(
            attempts = "3",
            backoff = @Backoff(delay = 1000, multiplier = 2.0),
            retryTopicSuffix = "#{'-' + '${spring.kafka.consumer.group-id}' + '-retry'}",
            dltTopicSuffix = "#{'-' + '${spring.kafka.consumer.group-id}' + '-dlt'}",
            numPartitions = "7"
    )
    @KafkaListener(topics = RETRYABLE_SINGLE_TOPIC, containerFactory = "retryableKafkaListenerContainerFactory")
//    @KafkaListener(topics = REPEATABLE_TOPIC)
    public void consumeRetryableSingle(Thing message) {
        log.info("Got message {}", message);
    }

    @KafkaListener(topics = RETRYABLE_DEFAULT_TOPIC, containerFactory = "retryableGlobalKafkaListenerContainerFactory")
    public void consumeRetryableDefault(Thing message) {
        log.info("Got message {}", message);
        throw new RuntimeException("Bad message");
    }
}

File: service-one/src/main/kotlin/com/dlt/kafkadlt/service/ConsumerService.kt
package com.dlt.kafkadlt.service

import com.dlt.kafkadlt.model.Info
import com.dlt.kafkadlt.model.Thing
import com.dlt.kafkadlt.utils.REPEATABLE_TOPIC
import com.dlt.kafkadlt.utils.TOPIC
import com.dlt.kafkadlt.utils.TOPIC_GOOD
import org.slf4j.LoggerFactory
import org.springframework.kafka.annotation.KafkaListener
import org.springframework.kafka.annotation.RetryableTopic
import org.springframework.kafka.retrytopic.TopicSuffixingStrategy
import org.springframework.retry.annotation.Backoff
import org.springframework.stereotype.Service
import java.lang.RuntimeException

@Service
class ConsumerService {
    companion object {
        val logger = LoggerFactory.getLogger(javaClass)

    }

    @KafkaListener(topics = [TOPIC], groupId = "group_id", errorHandler = "notificationListenerErrorHandler")
    fun consume(message: Info?) {
        logger.info("Got message $message")
        throw RuntimeException("Bad message")
    }

//    @KafkaListener(topics = [TOPIC_GOOD], groupId = "group_id", errorHandler = "notificationListenerErrorHandler")
    @KafkaListener(topics = [TOPIC_GOOD], groupId = "group_id")
    fun consumeGood(message: Thing?) {
        logger.info("Got message $message")
        throw RuntimeException("Bad message")
    }

    @RetryableTopic(
        attempts = "2",
        backoff = Backoff(delay = 1000, multiplier = 2.0),
        topicSuffixingStrategy = TopicSuffixingStrategy.SUFFIX_WITH_INDEX_VALUE,
        retryTopicSuffix = "group",
        dltTopicSuffix = "group-dlt",
        numPartitions = "7"
    )
    @KafkaListener(topics = [REPEATABLE_TOPIC], groupId = "group_id", containerFactory = "retryableKafkaListenerContainerFactory")
    fun consumeRepeatable(message: Thing?) {
        logger.info("Got message $message")
        throw RuntimeException("Bad message")
    }
}